common/migration/default-comparison.go

```diff
@@ -0,0 +1,157 @@
+		return false
+	}
+
+	// so, the problem here is that cmp.Diff doesn't love to receive pointers, and tends to panic
+	// for a fairly wide range of inputs when it hits a panic. This is not a test system, but rather
+	// an online flow, and so the use of a generic reflection tool is both slow and not particularly suited
+	// to this usecase, and quite dangerous insofar as that it panics a lot.
+	//
+	// That said, the ability to spot mapping differences, serialization differences and similar through its deep
+	// reflection is invaluable, so while this is a dangerous tradeoff, it's chosen here carefully, with the knowledge
+	// that this comparison is to be used in a context where panics are recovered and precautions such as the below are
+	// taken to try and catch the egregious cases which will cause panics.
+
+	// Firstly, use a bit of reflection to work out of the given types are pointers
+	// and try and derefrence them.
+	backgroundVal, d1, err := followPtrsAndReturnValue(backgroundRes)
+	if err != nil {
+		return false // fail to compare due to being unable to deref safely
+	}
+
+	activeVal, d2, err := followPtrsAndReturnValue(activeRes)
+	if err != nil {
+		return false // fail to compare due to being unable to deref safely
+	}
+
+	if d1 != d2 {
+		log.Warn("shadow flow got mismatched results, the results have a different number of pointers. Probably one of the values is a pointer and the other a value")
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	if backgroundVal == nil && activeVal == nil {
+		scope.IncCounter(metrics.ShadowerComparisonMatchCount)
+		return true
+	}
+
+	// and then go through and short-circuit if anything doesn't appear to be a struct
+	backgroundType := reflect.TypeOf(backgroundVal)
+	activeType := reflect.TypeOf(activeVal)
+
+	if backgroundType == nil || backgroundType.Kind() != reflect.Struct ||
+		activeType == nil || activeType.Kind() != reflect.Struct {
+
+		log.Warn("shadow flow got mismatched results, the types to compare are not struct or pointer to structs")
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+
+	diff := cmp.Diff(
+		backgroundVal,
+		activeVal,
+		cmpopts.IgnoreUnexported(backgroundVal),
+		cmpopts.IgnoreUnexported(activeVal))
+
+	if diff != "" {
+		log.Warn("mismatched results between active and background",
+			tag.Dynamic("comparison-result-res-diff", diff))
+		scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+		return false
+	}
+	scope.IncCounter(metrics.ShadowerComparisonMatchCount)
+	return true
+}
+
+func followPtrsAndReturnValue(in interface{}) (out interface{}, derefCount int, err error) {
+	derefCount = 0
+
+	out = in
+	for {
+		rv := reflect.ValueOf(out)
+		if rv.Kind() != reflect.Ptr {
+			break
+		}
+		derefCount++ // prevent infinite loops following pointers
+		if derefCount > 10 {
+			return nil, derefCount, fmt.Errorf("unable to find value, too many pointers")
```

Comment:
I should probably think about it harder, but I don't *think* it's possible to have a cycle in pointers like this.....

.... but I very much approve, 10 would be excessive enough to abort, and "no infinite loops ever" is entirely the correct choice in this kind of code.